frontend/src/components/editor/Editor.tsx

```diff
@@ -25,11 +28,173 @@ function Editor() {
 	const workspaceStore = useSelector(selectWorkspace);
 	const { mutateAsync: createUploadUrl } = useCreateUploadUrlMutation();
 	const { mutateAsync: uploadFile } = useUploadFileMutation();
+
+	const [formatBarState, setFormatBarState] = useState({
+		show: false,
+		position: { top: 0, left: 0 },
+		selectedFormats: new Set<FormatType>(),
+	});
+
 	const ref = useCallback((node: HTMLElement | null) => {
 		if (!node) return;
 		setElement(node);
 	}, []);
 
+	const getFormatMarker = useCallback((formatType: FormatType) => {
+		switch (formatType) {
+			case FormatType.BOLD:
+				return "**";
+			case FormatType.ITALIC:
+				return "_";
+			case FormatType.CODE:
+				return "`";
+			case FormatType.STRIKETHROUGH:
+				return "~~";
+		}
+	}, []);
+
+	const getFormatMarkerLength = useCallback((state: EditorState, from: number) => {
+		const maxCheckLength = 10;
+		const docSlice = state.sliceDoc(Math.max(0, from - maxCheckLength), from).toString();
+		let cnt = 0;
+
+		for (let i = docSlice.length - 1; i >= 0; i--) {
+			if (!["*", "_", "`", "~"].includes(docSlice[i])) break;
+			cnt++;
+		}
+
+		return cnt;
+	}, []);
+
+	const updateFormatBar = useCallback(
+		(update: ViewUpdate) => {
+			const selection = update.state.selection.main;
+			if (!selection.empty) {
+				const coords = update.view.coordsAtPos(selection.from);
+				if (coords) {
+					const maxLength = getFormatMarkerLength(update.view.state, selection.from);
+
+					const selectedTextStart = update.state.sliceDoc(
+						selection.from - maxLength,
+						selection.from
+					);
+					const selectedTextEnd = update.state.sliceDoc(
+						selection.to,
+						selection.to + maxLength
+					);
+					const formats = new Set<FormatType>();
+
+					const checkAndAddFormat = (marker: string, format: FormatType) => {
+						if (
+							selectedTextStart.includes(marker) &&
+							selectedTextEnd.includes(marker)
+						) {
+							formats.add(format);
+						}
+					};
+
+					checkAndAddFormat("**", FormatType.BOLD);
+					checkAndAddFormat("_", FormatType.ITALIC);
+					checkAndAddFormat("`", FormatType.CODE);
+					checkAndAddFormat("~~", FormatType.STRIKETHROUGH);
+
+					setFormatBarState((prev) => ({
+						...prev,
+						show: true,
+						position: {
+							top: coords.top - 8,
+							left: coords.left + 190,
+						},
+						selectedFormats: formats,
+					}));
+				}
+			} else {
+				setFormatBarState((prev) => ({
+					...prev,
+					show: false,
+					selectedFormats: new Set(),
+				}));
+			}
+		},
+		[getFormatMarkerLength]
+	);
+
+	const applyFormat = useCallback(
```

Comment:
How about separate codemirror shourcut extensions to other file?
It seems that we can move `applyFormat`, `getFormatMarker`, `getFormatMarkerLength`, `setKeymapConfig`,